### PR TITLE
[Jest] Migrate `image_source.test.js`

### PR DESCRIPTION
**Pull request by wipfli**
- Move and rename
- Run jscodeshift
- Run custom codemods and lint
- Manual edits

Hints on how to properly mock non-string XHR responses are welcome.

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Manually test the debug page.
 - [ ] Suggest a changelog category: bug/feature/docs/etc. or "skip changelog".
 - [ ] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Comment by birkskyum** [comment]
I think it's reasonable for us to use fetch at this point, if it makes it simpler in any way.

https://medium.com/swlh/how-to-mock-a-fetch-api-request-with-jest-and-typescript-bb6adf673a00
https://medium.com/swlh/mocking-xhr-request-with-jest-93875ed83dda

**Comment by birkskyum** [comment]
I meant both actually, and I also think we should keep it without using the network. I have tried to get the current "fires dataloading event" test to work, but for some reason the image isn't bound to the source.image, and it seems to be related to replacing window.useFakeXMLHttpRequest(); with our custom mock function.

**Comment by birkskyum** [comment]
I can see we still use sinon in the migrated style.test.ts, so it might also be an option to keep it.

**Comment by wipfli** [comment]
I got some inspiration from `load_glyph_range.test.ts`, still fails, but it feels closer.

**Comment by wipfli** [comment]
@HarelM do you think we need some `Object.defineProperty(global.Image.prototype, 'src'` stuff here, too?

**Comment by HarelM** [comment]
I haven't looked at it. It you want me to try and hack it, let me know...

**Comment by wipfli** [comment]
Please try and hack it, @HarelM :)

**Comment by wipfli** [comment]
Found something that works.

**Comment by birkskyum** [comment]
Awesome! And same approach as the style, glyph and ajax to keep things consistent :)

**Comment by birkskyum** [comment]
So is it ready for review?

**Comment by wipfli** [comment]
Thanks for looking at this. I will clean up and ping you when ready for review.

**Comment by wipfli** [comment]
This is the last pull request that migrates a unit test from tap to jest :) Thanks @HarelM, @astridx, @chippieTV, and @birkskyum for all the contributions :)

File: src/source/image_source.test.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+describe('ImageSource', () => {
+    const requests = [];
+    useFakeXMLHttpRequest().onCreate = (req) => { requests.push(req); };
+    beforeEach(() => {
+        global.fetch = null;
+    });
+
+    // mock createImageBitmap not supported
+    global.createImageBitmap = undefined;
```
Comment (HarelM) [inline src/source/image_source.test.ts:46]: Consider moving this code to utils or even better - stub ajax getImage method.

File: src/source/image_source.test.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        }
+    });
+
+    const respond = () => {
+        const req = requests.shift();
+        req.setStatus(200);
+        req.response = new ArrayBuffer(1);
+        req.onload();
+    };
+    jest.spyOn(browser, 'getImageData').mockImplementation(() => new ArrayBuffer(1) as any as ImageData);
```
Comment (HarelM) [inline src/source/image_source.test.ts:64]: I don't think this is needed. Browser is not used by image source anymore...

File: src/source/image_source.test.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        expect(source.maxzoom).toBe(22);
+        expect(source.tileSize).toBe(512);
+    });
+
+    test('fires dataloading event', () => {
+        const source = createSource({url : '/image.png'});
+        source.on('dataloading', (e) => {
+            expect(e.dataType).toBe('source');
+        });
+        source.onAdd(getStubMap());
```
Comment (HarelM) [inline src/source/image_source.test.ts:79]: I would keep it as it was as you are replacing a method with a single line of code. I would do `new MapStub...` as it was before.

**Review by HarelM — CHANGES_REQUESTED** [review]
Great work!! Small nits...

File: src/source/image_source.test.ts (hunk 1 above)
Comment (wipfli) [reply src/source/image_source.test.ts]: Thanks for reviewing. Did you think of something like https://github.com/maplibre/maplibre-gl-js/pull/732/commits/2a6bc1df7a351aedc69c7a5390039a55571b6d7f ?

Comment (HarelM) [reply src/source/image_source.test.ts]: This is good. Moving it to another file to be used by ajax would be better.
Stubbing ajax's getImage would be ideal :-)

**Review by HarelM — APPROVED** [review]

**Pull request merged**